cvat/apps/dataset_manager/task.py

```diff
@@ -115,47 +116,87 @@ def __init__(self, pk, is_prefetched=False):
     def reset(self):
         self.ir_data.reset()
 
+    def _validate_attribute_for_existence(self, db_attr_val, label_id, attr_type):
+        if db_attr_val.spec_id not in self.db_attributes[label_id][attr_type]:
+            raise AttributeError("spec_id `{}` is invalid".format(db_attr_val.spec_id))
+
+    def _validate_label_for_existence(self, label_id):
+        if label_id not in self.db_labels:
+            raise AttributeError("label_id `{}` is invalid".format(label_id))
+
+    def _update_parent_track(self, child_track, parent_track):
+        # parent track cannot have a frame greater than the frame of the child track
+        # so we need to prevent this situation
+        if parent_track is not None and child_track.frame < parent_track.frame:
+            parent_tracked_shapes = parent_track.trackedshape_set
+
+            if parent_tracked_shapes.count() == 1 and parent_tracked_shapes.first().type == "skeleton":
+                skeleton = parent_tracked_shapes.first()
+                skeleton.frame = child_track.frame
+                skeleton.save()
+
+            parent_track.frame = child_track.frame
+            parent_track.save()
+
+    def _add_missing_shape(self, track, shapes):
+        # first shape of track must have the same frame as track
+        first_shape = shapes[0]
+        for shape in shapes:
+            if shape["frame"] < first_shape["frame"]:
+                first_shape = shape
+
+        if track.frame < first_shape["frame"]:
+            missing_shape = deepcopy(first_shape)
+            missing_shape["frame"] = track.frame
+            missing_shape["outside"] = True
+            shapes.append(missing_shape)
+
+        elif track.frame > first_shape["frame"]:
+            track.frame = first_shape["frame"]
+
     def _save_tracks_to_db(self, tracks):
 
         def create_tracks(tracks, parent_track=None):
             db_tracks = []
-            db_track_attrvals = []
+            db_track_attr_vals = []
             db_shapes = []
-            db_shape_attrvals = []
+            db_shape_attr_vals = []
 
             for track in tracks:
                 track_attributes = track.pop("attributes", [])
                 shapes = track.pop("shapes")
                 elements = track.pop("elements", [])
                 db_track = models.LabeledTrack(job=self.db_job, parent=parent_track, **track)
-                if db_track.label_id not in self.db_labels:
-                    raise AttributeError("label_id `{}` is invalid".format(db_track.label_id))
+
+                self._validate_label_for_existence(db_track.label_id)
```

Comment:
If the checks are performed on saving, how the existing saved problems will be handled?

Reply:
This is not a new check, I just moved this code in a separate function. We don't have existing saved problems related to this check.

Reply:
If you mean another check that I added in this PR, you are right the existing save problems still exist. But in this case, it's possible to back up the task and upload it again. For sure we can write migration, but such migration will take a really long time.

Reply:
> If you mean another check that I added in this PR, you are right the existing save problems still exist. 

Ok, probably I've missed this.

> For sure we can write migration, but such migration will take a really long time.

Ok, but I thought about adding some checks during the export or track conversion to shapes instead. What do you thinks about this option?

Reply:
Yes, performing this check during the export also will resolve problem, but personally I don't like the idea of storing invalid annotations in db